build_tools/ci/amdxdna_driver_utils/amdxdna_ioctl.py

```diff
@@ -0,0 +1,217 @@
+
+IOC_NONE = 0
+IOC_WRITE = 1
+IOC_READ = 2
+
+
+def _IOC(dir, type, nr, size):
+    assert dir <= _IOC_DIRMASK, dir
+    assert type <= _IOC_TYPEMASK, type
+    assert nr <= _IOC_NRMASK, nr
+    assert size <= _IOC_SIZEMASK, size
+    return (
+        (dir << _IOC_DIRSHIFT)
+        | (type << _IOC_TYPESHIFT)
+        | (nr << _IOC_NRSHIFT)
+        | (size << _IOC_SIZESHIFT)
+    )
+
+
+def _IOC_TYPECHECK(t):
+    if isinstance(t, (memoryview, bytearray)):
+        size = len(t)
+    elif isinstance(t, struct.Struct):
+        size = t.size
+    elif isinstance(t, array.array):
+        size = t.itemsize * len(t)
+    else:
+        size = ctypes.sizeof(t)
+    assert size <= _IOC_SIZEMASK, size
+    return size
+
+
+def _IOWR(type, nr, size):
+    return _IOC(IOC_READ | IOC_WRITE, type, nr, _IOC_TYPECHECK(size))
+
+
+def get_struct(argp, stype):
+    return ctypes.cast(ctypes.c_void_p(argp), ctypes.POINTER(stype)).contents
+
+
+def get_void_ptr_to_struct(s):
+    ptr = ctypes.pointer(s)
+    return ctypes.cast(ptr, ctypes.c_void_p)
+
+
+def format_struct(s):
+    return pformat(s.as_dict(s))
+
+
+# <drm.h>
+DRM_IOCTL_BASE = ord("d")
+DRM_COMMAND_BASE = 0x40
+
+
+def DRM_IOWR(nr, type):
+    return _IOWR(DRM_IOCTL_BASE, nr, type)
+
+
+def ioctls_from_header():
+    hdr = (
+        (pathlib.Path(__file__).parent / "amdxdna_accel.py")
+        .read_text()
+        .replace("\\\n", "")
+    )
+    pattern = "DRM_IOCTL_AMDXDNA_([A-Z0-9_]+) = DRM_IOWR \( DRM_COMMAND_BASE \+ DRM_AMDXDNA_([A-Z0-9_]+) , struct_amdxdna_drm_([a-z0-9_]+) \)"
+    matches = re.findall(pattern, hdr, re.MULTILINE)
+    ioctls = Namespace()
+    for name, offset, sname in matches:
+        assert name == offset
+        offset = f"DRM_AMDXDNA_{name}"
+        assert hasattr(amdxdna_accel, offset)
+        offset = getattr(amdxdna_accel, offset)
+        struc = getattr(amdxdna_accel, "struct_amdxdna_drm_" + sname)
+        setattr(
+            ioctls,
+            f"DRM_IOCTL_AMDXDNA_{name}",
+            DRM_IOWR(DRM_COMMAND_BASE + offset, struc),
+        )
+
+    return ioctls
+
+
+ioctls = ioctls_from_header()
+
+
+def get_aie_version(drv_fd):
+    version = struct_amdxdna_drm_query_aie_version()
+    info_params = struct_amdxdna_drm_get_info(
+        DRM_AMDXDNA_QUERY_AIE_VERSION,
+        ctypes.sizeof(struct_amdxdna_drm_query_aie_version),
+        get_void_ptr_to_struct(version).value,
+    )
+
+    fcntl.ioctl(drv_fd, ioctls.DRM_IOCTL_AMDXDNA_GET_INFO, info_params)
+
+    return version.major, version.minor
+
+
+def get_aie_metadata(drv_fd):
+    metadata = struct_amdxdna_drm_query_aie_metadata()
+    info_params = struct_amdxdna_drm_get_info(
+        DRM_AMDXDNA_QUERY_AIE_METADATA,
+        ctypes.sizeof(struct_amdxdna_drm_query_aie_metadata),
+        get_void_ptr_to_struct(metadata).value,
+    )
+
+    fcntl.ioctl(drv_fd, ioctls.DRM_IOCTL_AMDXDNA_GET_INFO, info_params)
+
+    return format_struct(metadata)
+
+
+def get_core_n_rows(drv_fd):
```

Comment:
Why can't we retrieve this in the HAL driver itself?

Reply:
You can - I have helper methods in [shim/linux/kmq/device.cpp](https://github.com/nod-ai/iree-amd-aie/blob/fc364b727c72050c20aceaaaaff3550a94a3c4e9/runtime/src/iree-amd-aie/driver/xrt-lite/shim/linux/kmq/device.cpp#L252-L275) that do the exact same things but I had the feeling that, since this is core config data, it should be passed in. I.e., you might want to override at some point (for some reason).